File: scripts/generate_icons.py
from pathlib import Path
from collections import deque

from PIL import Image

ROOT = Path(__file__).resolve().parents[1] / "public" / "icons"
SOURCE = ROOT / "recipe-app-icon.png"


def remove_outer_white_background(image: Image.Image) -> Image.Image:
    """Make only edge-connected near-white pixels transparent.

    White details inside the illustration stay intact; the cream rounded border
    is below the near-white threshold and remains opaque.
    """
    image = image.convert("RGBA")
    pixels = image.load()
    width, height = image.size
    queued = deque()
    visited = set()

    def is_outer_white(x: int, y: int) -> bool:
        red, green, blue, _ = pixels[x, y]
        return red >= 250 and green >= 250 and blue >= 250

    for x in range(width):
        queued.extend(((x, 0), (x, height - 1)))
    for y in range(height):
        queued.extend(((0, y), (width - 1, y)))

    while queued:
        x, y = queued.popleft()
        if (x, y) in visited or not is_outer_white(x, y):
            continue
        visited.add((x, y))
        red, green, blue, _ = pixels[x, y]
        pixels[x, y] = (red, green, blue, 0)
        for nx, ny in ((x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)):
            if 0 <= nx < width and 0 <= ny < height:
                queued.append((nx, ny))
    return image


def make_icon(size: int, filename: str) -> None:
    image = remove_outer_white_background(Image.open(SOURCE))
    image.thumbnail((size, size), Image.Resampling.LANCZOS)
    image.save(ROOT / filename, optimize=True)


make_icon(192, "icon-192.png")
make_icon(512, "icon-512.png")
make_icon(180, "apple-touch-icon.png")

File: scripts/generate_icons.mjs
import { resolve } from 'node:path';
import sharp from 'sharp';

const icons = resolve('public/icons');
const source = resolve(icons, 'recipe-app-icon.png');

const { data, info } = await sharp(source).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
const queue = [];
const visited = new Uint8Array(info.width * info.height);

const isOuterWhite = (x, y) => {
  const offset = (y * info.width + x) * 4;
  return data[offset] >= 250 && data[offset + 1] >= 250 && data[offset + 2] >= 250;
};

for (let x = 0; x < info.width; x += 1) queue.push([x, 0], [x, info.height - 1]);
for (let y = 0; y < info.height; y += 1) queue.push([0, y], [info.width - 1, y]);

for (let index = 0; index < queue.length; index += 1) {
  const [x, y] = queue[index];
  const pixel = y * info.width + x;
  if (visited[pixel] || !isOuterWhite(x, y)) continue;
  visited[pixel] = 1;
  data[pixel * 4 + 3] = 0;
  if (x > 0) queue.push([x - 1, y]);
  if (x + 1 < info.width) queue.push([x + 1, y]);
  if (y > 0) queue.push([x, y - 1]);
  if (y + 1 < info.height) queue.push([x, y + 1]);
}

const cleaned = sharp(data, { raw: info });
await Promise.all([
  cleaned.clone().resize(192, 192).png().toFile(resolve(icons, 'icon-192.png')),
  cleaned.clone().resize(512, 512).png().toFile(resolve(icons, 'icon-512.png')),
  cleaned.clone().resize(180, 180).png().toFile(resolve(icons, 'apple-touch-icon.png')),
]);
